File: computing/interview/oop/csharp/ShapeExample.cs
using System;
using System.Collections.Generic;

namespace Interview.Oop
{
    public abstract class Shape
    {
        public abstract double Area { get; }
        public abstract void Describe();
    }

    public sealed class Circle : Shape
    {
        private readonly double _radius;

        public Circle(double radius)
        {
            _radius = radius;
        }

        public override double Area => Math.PI * _radius * _radius;

        public override void Describe()
        {
            Console.WriteLine($"Circle radius={_radius} area={Area:0.###}");
        }
    }

    public sealed class Rectangle : Shape
    {
        private readonly double _width;
        private readonly double _height;

        public Rectangle(double width, double height)
        {
            _width = width;
            _height = height;
        }

        public override double Area => _width * _height;

        public override void Describe()
        {
            Console.WriteLine($"Rectangle {_width}x{_height} area={Area:0.###}");
        }
    }

    public static class Program
    {
        public static void Main()
        {
            var shapes = new List<Shape>
            {
                new Circle(2.0),
                new Rectangle(3.0, 4.0)
            };

            foreach (var shape in shapes)
            {
                shape.Describe();
            }
        }
    }
}

File: computing/interview/oop/java/ShapeExample.java
package interview.oop;

import java.util.ArrayList;
import java.util.List;

abstract class Shape {
    abstract double area();
    abstract void describe();
}

final class Circle extends Shape {
    private final double radius;

    Circle(double radius) {
        this.radius = radius;
    }

    @Override
    double area() {
        return Math.PI * radius * radius;
    }

    @Override
    void describe() {
        System.out.printf("Circle radius=%.2f area=%.3f%n", radius, area());
    }
}

final class Rectangle extends Shape {
    private final double width;
    private final double height;

    Rectangle(double width, double height) {
        this.width = width;
        this.height = height;
    }

    @Override
    double area() {
        return width * height;
    }

    @Override
    void describe() {
        System.out.printf("Rectangle %.2fx%.2f area=%.3f%n", width, height, area());
    }
}

public final class ShapeExample {
    private ShapeExample() {}

    public static void main(String[] args) {
        List<Shape> shapes = new ArrayList<>();
        shapes.add(new Circle(2.0));
        shapes.add(new Rectangle(3.0, 4.0));

        for (Shape shape : shapes) {
            shape.describe();
        }
    }
}
